### ISSUE TITLE
Bug: App exits unexpectedly when run with appmap agent

### ISSUE THREAD

**Issue by ptrdvrk** [applandinc/appmap-agent-js#36]
App: https://github.com/sasha7/notes-app

Console output, see attached debug log
```
GET / 304 26.244 ms - -
GET /css/vendor/bootstrap/css/bootstrap.min.css 304 4.308 ms - -
GET /css/flat-ui.css 304 3.216 ms - -
GET /js/application.js 304 1.989 ms - -
GET /css/style.css 304 2.103 ms - -
GET /js/vendor/jquery.min.js 304 3.042 ms - -
GET /js/flat-ui.min.js 304 3.036 ms - -
GET /fonts/lato/lato-bold.woff 304 1.144 ms - -
GET /fonts/lato/lato-light.woff 304 1.033 ms - -
GET /fonts/lato/lato-regular.woff 304 0.938 ms - -
GET /img/icons/svg/clipboard.svg 404 11.159 ms - 4277
received event after closing both request and response
APPMAP-INFO > Exited with: 1
```
[log.txt](https://github.com/applandinc/appmap-agent-js/files/7429958/log.txt)

**Comment by lachrist** [comment]
This should be fixed by: https://github.com/applandinc/appmap-agent-js/commit/bbc3aca85f6ffc9df7697f5d1f1a211c81091e38

Could you check that is solved in v6.3.1?

**Comment by ptrdvrk** [comment]
@lachrist I am blocked by the same issue as in #37: https://github.com/applandinc/appmap-agent-js/issues/37#issuecomment-964457575

**Comment by ptrdvrk** [comment]
@lachrist still blocked in 8.0.0, log below FYI.

```
cat appmap.yml
name: Note
packages:
  - path: models
  - path: routes
  - path: views
  - path: controllers
  - path: lib

‣ npx appmap-agent-js --version
@appland/appmap-agent-js@8.0.0

‣ npx appmap-agent-js -- node ./app.js
APPMAP-WARNING command "git describe --abbrev=0 --tags" on cwd "/Users/ptrdvrk/work/github/js/notes-app" failed with 128 >> fatal: No names found, cannot describe anything.

APPMAP-WARNING command "git describe --abbrev=0" on cwd "/Users/ptrdvrk/work/github/js/notes-app" failed with 128 >> fatal: No names found, cannot describe anything.

APPMAP-WARNING command "git describe --long --tags" on cwd "/Users/ptrdvrk/work/github/js/notes-app" failed with 128 >> fatal: No names found, cannot describe anything.

APPMAP-WARNING command "git describe --long" on cwd "/Users/ptrdvrk/work/github/js/notes-app" failed with 128 >> fatal: No names found, cannot describe anything.

APPMAP-INFO Trace port: 50208
APPMAP-INFO Track port: 50209
APPMAP-INFO 'node' './app.js' ...
(node:76986) ExperimentalWarning: --experimental-loader is an experimental feature. This feature could change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
APPMAP-INFO Please, ignore node's deprecated warning about outdated transformSource loader hook (if present).
APPMAP-INFO Recorder 'remote' caught process 76986
APPMAP-INFO Recording /Users/ptrdvrk/work/github/js/notes-app/app.js.
APPMAP-WARNING Data url for source map is encoded as base64 and does not declare UTF-8 as its character encoding, will try to use UTF-8 anyway >> application/json;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbIm9iamVjdGlvbi5qcyJdLCJuYW1lcyI6WyJNb2RlbEJhc2UiLCJNb2RlbCIsIlF1ZXJ5QnVpbGRlciIsIlF1ZXJ5QnVpbGRlckJhc2UiLCJRdWVyeUJ1aWxkZXJPcGVyYXRpb24iLCJSZWxhdGlvbkV4cHJlc3Npb24iLCJWYWxpZGF0aW9uRXJyb3IiLCJSZWxhdGlvbiIsIkhhc09uZVJlbGF0aW9uIiwiSGFzTWFueVJlbGF0aW9uIiwiQmVsb25nc1RvT25lUmVsYXRpb24iLCJNYW55VG9NYW55UmVsYXRpb24iLCJ0cmFuc2FjdGlvbiIsIlByb21pc2UiLCJPYmplY3QiLCJkZWZpbmVQcm9wZXJ0eSIsIm1vZHVsZSIsImV4cG9ydHMiLCJnZXQiLCJjb25zb2xlIiwid2FybiJdLCJtYXBwaW5ncyI6Ijs7Ozs7OztBQUFBOzs7O0FBQ0E7Ozs7QUFDQTs7OztBQUNBOzs7O0FBQ0E7Ozs7QUFDQTs7OztBQUNBOzs7O0FBRUE7Ozs7QUFDQTs7OztBQUNBOzs7O0FBQ0E7Ozs7QUFDQTs7OztBQUVBOzs7O0FBQ0E7Ozs7OztRQUdFQSxTO1FBQ0FDLEs7UUFDQUMsWTtRQUNBQyxnQjtRQUNBQyxxQjtRQUNBQyxrQjtRQUNBQyxlO1FBQ0FDLFE7UUFDQUMsYztRQUNBQyxlO1FBQ0FDLG9CO1FBQ0FDLGtCO1FBQ0FDLFc7UUFDQUMsTzs7O0FBR0ZDLE9BQU9DLGNBQVAsQ0FBc0JDLE9BQU9DLE9BQTdCLEVBQXNDLGtCQUF0QyxFQUEwRDtBQUN4REMsT0FBSyxlQUFZO0FBQ2ZDLFlBQVFDLElBQVI7QUFDQTtBQUNEO0FBSnVELENBQTFEOztBQU9BTixPQUFPQyxjQUFQLENBQXNCQyxPQUFPQyxPQUE3QixFQUFzQyxtQkFBdEMsRUFBMkQ7QUFDekRDLE9BQUssZUFBWTtBQUNmQyxZQUFRQyxJQUFSO0FBQ0E7QUFDRDtBQUp3RCxDQUEzRCIsImZpbGUiOiJvYmplY3Rpb24uanMiLCJzb3VyY2VzQ29udGVudCI6WyJpbXBvcnQgTW9kZWxCYXNlIGZyb20gJy4vbW9kZWwvTW9kZWxCYXNlJztcbmltcG9ydCBNb2RlbCBmcm9tICcuL21vZGVsL01vZGVsJztcbmltcG9ydCBRdWVyeUJ1aWxkZXJCYXNlIGZyb20gJy4vcXVlcnlCdWlsZGVyL1F1ZXJ5QnVpbGRlckJhc2UnO1xuaW1wb3J0IFF1ZXJ5QnVpbGRlciBmcm9tICcuL3F1ZXJ5QnVpbGRlci9RdWVyeUJ1aWxkZXInO1xuaW1wb3J0IFF1ZXJ5QnVpbGRlck9wZXJhdGlvbiBmcm9tICcuL3F1ZXJ5QnVpbGRlci9vcGVyYXRpb25zL1F1ZXJ5QnVpbGRlck9wZXJhdGlvbidcbmltcG9ydCBSZWxhdGlvbkV4cHJlc3Npb24gZnJvbSAnLi9xdWVyeUJ1aWxkZXIvUmVsYXRpb25FeHByZXNzaW9uJztcbmltcG9ydCBWYWxpZGF0aW9uRXJyb3IgZnJvbSAnLi9WYWxpZGF0aW9uRXJyb3InO1xuXG5pbXBvcnQgUmVsYXRpb24gZnJvbSAnLi9yZWxhdGlvbnMvUmVsYXRpb24nO1xuaW1wb3J0IEhhc09uZVJlbGF0aW9uIGZyb20gJy4vcmVsYXRpb25zL2hhc09uZS9IYXNPbmVSZWxhdGlvbic7XG5pbXBvcnQgSGFzTWFueVJlbGF0aW9uIGZyb20gJy4vcmVsYXRpb25zL2hhc01hbnkvSGFzTWFueVJlbGF0aW9uJztcbmltcG9ydCBCZWxvbmdzVG9PbmVSZWxhdGlvbiBmcm9tICcuL3JlbGF0aW9ucy9iZWxvbmdzVG9PbmUvQmVsb25nc1RvT25lUmVsYXRpb24nO1xuaW1wb3J0IE1hbnlUb01hbnlSZWxhdGlvbiBmcm9tICcuL3JlbGF0aW9ucy9tYW55VG9NYW55L01hbnlUb01hbnlSZWxhdGlvbic7XG5cbmltcG9ydCB0cmFuc2FjdGlvbiBmcm9tICcuL3RyYW5zYWN0aW9uJztcbmltcG9ydCBQcm9taXNlIGZyb20gJ2JsdWViaXJkJztcblxuZXhwb3J0IHtcbiAgTW9kZWxCYXNlLFxuICBNb2RlbCxcbiAgUXVlcnlCdWlsZGVyLFxuICBRdWVyeUJ1aWxkZXJCYXNlLFxuICBRdWVyeUJ1aWxkZXJPcGVyYXRpb24sXG4gIFJlbGF0aW9uRXhwcmVzc2lvbixcbiAgVmFsaWRhdGlvbkVycm9yLFxuICBSZWxhdGlvbixcbiAgSGFzT25lUmVsYXRpb24sXG4gIEhhc01hbnlSZWxhdGlvbixcbiAgQmVsb25nc1RvT25lUmVsYXRpb24sXG4gIE1hbnlUb01hbnlSZWxhdGlvbixcbiAgdHJhbnNhY3Rpb24sXG4gIFByb21pc2Vcbn07XG5cbk9iamVjdC5kZWZpbmVQcm9wZXJ0eShtb2R1bGUuZXhwb3J0cywgXCJPbmVUb09uZVJlbGF0aW9uXCIsIHtcbiAgZ2V0OiBmdW5jdGlvbiAoKSB7XG4gICAgY29uc29sZS53YXJuKGBPbmVUb09uZVJlbGF0aW9uIGlzIGRlcHJlY2F0ZWQgYW5kIHdpbGwgYmUgcmVtb3ZlZCBpbiB2ZXJzaW9uIDAuNy4wLiBVc2UgQmVsb25nc1RvT25lUmVsYXRpb24gaW5zdGVhZC4gU2ltcGx5IHJlcGxhY2UgT25lVG9PbmVSZWxhdGlvbiB3aXRoIEJlbG9uZ3NUb09uZVJlbGF0aW9uLmApO1xuICAgIHJldHVybiBCZWxvbmdzVG9PbmVSZWxhdGlvbjtcbiAgfVxufSk7XG5cbk9iamVjdC5kZWZpbmVQcm9wZXJ0eShtb2R1bGUuZXhwb3J0cywgXCJPbmVUb01hbnlSZWxhdGlvblwiLCB7XG4gIGdldDogZnVuY3Rpb24gKCkge1xuICAgIGNvbnNvbGUud2FybihgT25lVG9NYW55UmVsYXRpb24gaXMgZGVwcmVjYXRlZCBhbmQgd2lsbCBiZSByZW1vdmVkIGluIHZlcnNpb24gMC43LjAuIFVzZSBIYXNNYW55UmVsYXRpb24gaW5zdGVhZC4gU2ltcGx5IHJlcGxhY2UgT25lVG9NYW55UmVsYXRpb24gd2l0aCBIYXNNYW55UmVsYXRpb24uYCk7XG4gICAgcmV0dXJuIEhhc01hbnlSZWxhdGlvbjtcbiAgfVxufSk7XG5cbiJdfQ==
Expected an absolute path because the reference url is a data url, got "objection.js" relative to "data:application/json;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbIm9iamVjdGlvbi5qcyJdLCJuYW1lcyI6WyJNb2RlbEJhc2UiLCJNb2RlbCIsIlF1ZXJ5QnVpbGRlciIsIlF1ZXJ5QnVpbGRlckJhc2UiLCJRdWVyeUJ1aWxkZXJPcGVyYXRpb24iLCJSZWxhdGlvbkV4cHJlc3Npb24iLCJWYWxpZGF0aW9uRXJyb3IiLCJSZWxhdGlvbiIsIkhhc09uZVJlbGF0aW9uIiwiSGFzTWFueVJlbGF0aW9uIiwiQmVsb25nc1RvT25lUmVsYXRpb24iLCJNYW55VG9NYW55UmVsYXRpb24iLCJ0cmFuc2FjdGlvbiIsIlByb21pc2UiLCJPYmplY3QiLCJkZWZpbmVQcm9wZXJ0eSIsIm1vZHVsZSIsImV4cG9ydHMiLCJnZXQiLCJjb25zb2xlIiwid2FybiJdLCJtYXBwaW5ncyI6Ijs7Ozs7OztBQUFBOzs7O0FBQ0E7Ozs7QUFDQTs7OztBQUNBOzs7O0FBQ0E7Ozs7QUFDQTs7OztBQUNBOzs7O0FBRUE7Ozs7QUFDQTs7OztBQUNBOzs7O0FBQ0E7Ozs7QUFDQTs7OztBQUVBOzs7O0FBQ0E7Ozs7OztRQUdFQSxTO1FBQ0FDLEs7UUFDQUMsWTtRQUNBQyxnQjtRQUNBQyxxQjtRQUNBQyxrQjtRQUNBQyxlO1FBQ0FDLFE7UUFDQUMsYztRQUNBQyxlO1FBQ0FDLG9CO1FBQ0FDLGtCO1FBQ0FDLFc7UUFDQUMsTzs7O0FBR0ZDLE9BQU9DLGNBQVAsQ0FBc0JDLE9BQU9DLE9BQTdCLEVBQXNDLGtCQUF0QyxFQUEwRDtBQUN4REMsT0FBSyxlQUFZO0FBQ2ZDLFlBQVFDLElBQVI7QUFDQTtBQUNEO0FBSnVELENBQTFEOztBQU9BTixPQUFPQyxjQUFQLENBQXNCQyxPQUFPQyxPQUE3QixFQUFzQyxtQkFBdEMsRUFBMkQ7QUFDekRDLE9BQUssZUFBWTtBQUNmQyxZQUFRQyxJQUFSO0FBQ0E7QUFDRDtBQUp3RCxDQUEzRCIsImZpbGUiOiJvYmplY3Rpb24uanMiLCJzb3VyY2VzQ29udGVudCI6WyJpbXBvcnQgTW9kZWxCYXNlIGZyb20gJy4vbW9kZWwvTW9kZWxCYXNlJztcbmltcG9ydCBNb2RlbCBmcm9tICcuL21vZGVsL01vZGVsJztcbmltcG9ydCBRdWVyeUJ1aWxkZXJCYXNlIGZyb20gJy4vcXVlcnlCdWlsZGVyL1F1ZXJ5QnVpbGRlckJhc2UnO1xuaW1wb3J0IFF1ZXJ5QnVpbGRlciBmcm9tICcuL3F1ZXJ5QnVpbGRlci9RdWVyeUJ1aWxkZXInO1xuaW1wb3J0IFF1ZXJ5QnVpbGRlck9wZXJhdGlvbiBmcm9tICcuL3F1ZXJ5QnVpbGRlci9vcGVyYXRpb25zL1F1ZXJ5QnVpbGRlck9wZXJhdGlvbidcbmltcG9ydCBSZWxhdGlvbkV4cHJlc3Npb24gZnJvbSAnLi9xdWVyeUJ1aWxkZXIvUmVsYXRpb25FeHByZXNzaW9uJztcbmltcG9ydCBWYWxpZGF0aW9uRXJyb3IgZnJvbSAnLi9WYWxpZGF0aW9uRXJyb3InO1xuXG5pbXBvcnQgUmVsYXRpb24gZnJvbSAnLi9yZWxhdGlvbnMvUmVsYXRpb24nO1xuaW1wb3J0IEhhc09uZVJlbGF0aW9uIGZyb20gJy4vcmVsYXRpb25zL2hhc09uZS9IYXNPbmVSZWxhdGlvbic7XG5pbXBvcnQgSGFzTWFueVJlbGF0aW9uIGZyb20gJy4vcmVsYXRpb25zL2hhc01hbnkvSGFzTWFueVJlbGF0aW9uJztcbmltcG9ydCBCZWxvbmdzVG9PbmVSZWxhdGlvbiBmcm9tICcuL3JlbGF0aW9ucy9iZWxvbmdzVG9PbmUvQmVsb25nc1RvT25lUmVsYXRpb24nO1xuaW1wb3J0IE1hbnlUb01hbnlSZWxhdGlvbiBmcm9tICcuL3JlbGF0aW9ucy9tYW55VG9NYW55L01hbnlUb01hbnlSZWxhdGlvbic7XG5cbmltcG9ydCB0cmFuc2FjdGlvbiBmcm9tICcuL3RyYW5zYWN0aW9uJztcbmltcG9ydCBQcm9taXNlIGZyb20gJ2JsdWViaXJkJztcblxuZXhwb3J0IHtcbiAgTW9kZWxCYXNlLFxuICBNb2RlbCxcbiAgUXVlcnlCdWlsZGVyLFxuICBRdWVyeUJ1aWxkZXJCYXNlLFxuICBRdWVyeUJ1aWxkZXJPcGVyYXRpb24sXG4gIFJlbGF0aW9uRXhwcmVzc2lvbixcbiAgVmFsaWRhdGlvbkVycm9yLFxuICBSZWxhdGlvbixcbiAgSGFzT25lUmVsYXRpb24sXG4gIEhhc01hbnlSZWxhdGlvbixcbiAgQmVsb25nc1RvT25lUmVsYXRpb24sXG4gIE1hbnlUb01hbnlSZWxhdGlvbixcbiAgdHJhbnNhY3Rpb24sXG4gIFByb21pc2Vcbn07XG5cbk9iamVjdC5kZWZpbmVQcm9wZXJ0eShtb2R1bGUuZXhwb3J0cywgXCJPbmVUb09uZVJlbGF0aW9uXCIsIHtcbiAgZ2V0OiBmdW5jdGlvbiAoKSB7XG4gICAgY29uc29sZS53YXJuKGBPbmVUb09uZVJlbGF0aW9uIGlzIGRlcHJlY2F0ZWQgYW5kIHdpbGwgYmUgcmVtb3ZlZCBpbiB2ZXJzaW9uIDAuNy4wLiBVc2UgQmVsb25nc1RvT25lUmVsYXRpb24gaW5zdGVhZC4gU2ltcGx5IHJlcGxhY2UgT25lVG9PbmVSZWxhdGlvbiB3aXRoIEJlbG9uZ3NUb09uZVJlbGF0aW9uLmApO1xuICAgIHJldHVybiBCZWxvbmdzVG9PbmVSZWxhdGlvbjtcbiAgfVxufSk7XG5cbk9iamVjdC5kZWZpbmVQcm9wZXJ0eShtb2R1bGUuZXhwb3J0cywgXCJPbmVUb01hbnlSZWxhdGlvblwiLCB7XG4gIGdldDogZnVuY3Rpb24gKCkge1xuICAgIGNvbnNvbGUud2FybihgT25lVG9NYW55UmVsYXRpb24gaXMgZGVwcmVjYXRlZCBhbmQgd2lsbCBiZSByZW1vdmVkIGluIHZlcnNpb24gMC43LjAuIFVzZSBIYXNNYW55UmVsYXRpb24gaW5zdGVhZC4gU2ltcGx5IHJlcGxhY2UgT25lVG9NYW55UmVsYXRpb24gd2l0aCBIYXNNYW55UmVsYXRpb24uYCk7XG4gICAgcmV0dXJuIEhhc01hbnlSZWxhdGlvbjtcbiAgfVxufSk7XG5cbiJdfQ=="
APPMAP-INFO > Exited with: 1
```

**Comment by ptrdvrk** [comment]
I no longer see the hard crashes, good! The warnings are still present, is this a benign state or shall we open a new ticket for this @lachrist ?
```
POST /login 302 60020.566 ms - 56
GET /login 200 17.022 ms - 2418
GET /css/vendor/bootstrap/css/bootstrap.min.css 304 0.747 ms - -
GET /css/flat-ui.css 304 1.369 ms - -
GET /css/style.css 304 1.102 ms - -
GET /js/vendor/jquery.min.js 304 2.646 ms - -
GET /js/flat-ui.min.js 304 0.577 ms - -
GET /fonts/lato/lato-regular.woff 304 0.720 ms - -
GET /fonts/glyphicons/flat-ui-icons-regular.woff 304 0.534 ms - -
GET /img/icons/svg/clipboard.svg 404 16.678 ms - 4277
APPMAP-WARNING HTTP activity detected after closing both the request and the response.
```

**Comment by lachrist** [comment]
The agent assumed that an http cycle was done yet it detected some lingering activity afterward. I can imagine the following reasons for it:

1) The method for detecting the end of the cycle is wrong. This means that we are missing some information in the bundle of events associated to the cycle. I've tried different method and the current seems to work the best (least amount of warnings).
2) The node engine recycled the request/response. This is more serious because we might miss entire cycles. I'm not aware that node would do that but who knows...

I think this is one of these cases where if nothing seems to be wrong then we can just leave this issue at rest. Yet we should know that is a potential vulnerability in the agent.

**Comment by ptrdvrk** [comment]
Sounds like we'll have ticket for this in the backlock and keep an eye on this problem going forward.
